libvips/foreign/vips2tiff.c

```diff
@@ -629,6 +643,167 @@ wtiff_embed_imagedescription(Wtiff *wtiff, TIFF *tif)
 	return 0;
 }
 
+#ifdef HAVE_JPEG
+// in vips2jpeg.c
+void vips__jpeg_target_dest(j_compress_ptr cinfo, VipsTarget *target);
+
+static void
+wtiff_compress_jpeg_header(Wtiff *wtiff,
+	struct jpeg_compress_struct *cinfo, VipsImage *image)
+{
+	J_COLOR_SPACE space;
+
+	cinfo->image_width = wtiff->tilew;
+	cinfo->image_height = wtiff->tileh;
+	cinfo->input_components = image->Bands;
+
+	if (image->Bands == 4 &&
+		image->Type == VIPS_INTERPRETATION_CMYK) {
+		space = JCS_CMYK;
+	}
+	else if (image->Bands == 3)
+		space = JCS_RGB;
+	else if (image->Bands == 1)
+		space = JCS_GRAYSCALE;
+	else
+		/* Use luminance compression for all channels.
+		 */
+		space = JCS_UNKNOWN;
+	cinfo->in_color_space = space;
+
+#ifdef HAVE_JPEG_EXT_PARAMS
+	/* Reset compression profile to libjpeg defaults
+	 */
+	if (jpeg_c_int_param_supported(cinfo, JINT_COMPRESS_PROFILE))
+		jpeg_c_set_int_param(cinfo,
+			JINT_COMPRESS_PROFILE, JCP_FASTEST);
+#endif
+
+	jpeg_set_defaults(cinfo);
+
+	/* Set compression quality. Must be called after setting params above.
+	 */
+	jpeg_set_quality(cinfo, wtiff->Q, TRUE);
+
+	// disable chroma subsample for high Q
+	if (wtiff->Q >= 90) {
+		int i;
+
+		for (i = 0; i < image->Bands; i++) {
+			cinfo->comp_info[i].h_samp_factor = 1;
+			cinfo->comp_info[i].v_samp_factor = 1;
+		}
+	}
```

Comment:
When I omit the JFIF APP0 marker, I get file sizes similar to those of the Git master.
```suggestion
	}

	// Avoid writing the JFIF APP0 marker.
	cinfo->write_JFIF_header = FALSE;
```

Reply:
Ah nice! I forgot about this.